Initial implementation of Hollowverse Architecture website

wholesomedev opened the pull request:
~~This still needs more work, but a demo has been deployed to: https://hollowverse.github.io/architecture~~

~~To test:~~

* ~~Click on "Visitor request processing"~~
* ~~Click on "Visitor", "Browser", or click on the link between them to see the currently available interactive features~~

~~The plan is to list the architectures that will be documented in the left sidebar. Currently the only listed architecture is "Visitor request processing" (which I should probably rename to just "Visitor request"), but there are more that we should list, such as our deployment architecture, Hollowverse GitHub codebase architecture, and others.~~

~~For the "Visitor request" architecture, there's obviously many more components than just "Visitor" and "Browser", like Route 53, CloudFront, Lambdas, etc. I frankly don't know all the involved components or how they interact with each other. I'll need to discuss with @forabi to find out and document.~~

~~We also need to figure out where and how we'll deploy this website. GitHub pages seem to be pretty limited. (Not sure if it can support deep links.)~~

## ~~To-dos:~~

- [x] ~~Fix TODO comments in the code~~
- [x] ~~Remove Redux and its related dependencies (reduce complexity)~~
- [x] ~~Clean up webpack configurations (no need for `hollowverse/hollowverse`-level complexity here)~~
- [x] ~~Finish the "Visitor request" architecture~~
- [x] ~~Figure out deployment~~
- [x] ~~Clean up `package.json`~~
- [x] ~~Remove unnecessary files~~

Closes https://github.com/hollowverse/hollowverse/issues/244

wholesomedev committed:
Add initial clown configs

wholesomedev committed:
Add some components

wholesomedev committed:
Add files from hollowverse/hollowverse with some clean up

wholesomedev committed:
Remove more unnecessary code from hollowverse/hollowverse

wholesomedev committed:
Get the app to compile after a bunch of clean up

wholesomedev committed:
Add some basic layout

wholesomedev committed:
Add some architectural components and their documentation

wholesomedev committed:
Add some layout elements

wholesomedev committed:
Add the first architecture

wholesomedev committed:
Add Redux boilerplate for setting the selected architecture component

wholesomedev committed:
Add `launch.json` to debug in Chrome

wholesomedev committed:
Move components to their own files, create Architecture and AcDocumentation components

wholesomedev committed:
create componentMap and linkMap

wholesomedev committed:
Refactor code a bit

wholesomedev committed:
Get edge selection to work

wholesomedev committed:
Make TypeScript happy

wholesomedev committed:
Fix build

wholesomedev committed:
Make lint happy

forabi commented:
Shouldn't we use some open-source documentation tool? I feel like this is going to take a considerable effort to build and maintain.

> (Not sure if it can support deep links.)

Can't we use React Router with [`HashRouter`](https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/docs/api/HashRouter.md) or hash history for deep linking?

>  I'll need to discuss with @forabi to find out and document.

So here is what happens when a user request a page:

1. Browsers resolves DNS for hollowverse.com. AWS Route 53 manages our DNS records.
2. hollowvere.com is configured as an alias for a CloudFront distribution domain name assigned by AWS.
  
  ![screenshot from 2018-04-29 19-21-51](https://user-images.githubusercontent.com/1012761/39408682-98e67912-4be2-11e8-9cd3-c30862377dd2.png)
  
3. The DNS request resolves to one of the IPs of the the CF distribution. Ideally, this would be the IP of the CF server that's closest to the user.
4. Now CF receives the request for the page:
   1. CF executes the Lambda@Edge assigned to the `viewer-request` stage. This is [`assignEnvironmentToViewerRequest`](https://github.com/hollowverse/route-request/blob/master/src/assignEnvironmentToViewerRequest.ts).
   2. `assignEnvironmentToViewerRequest` modifies the incoming user request and returns the modified request to CF. CF now looks for a cached copy of the page that matches exactly the headers the constitute the cache key for a specific page. It's important to note that this check happens _after_ `assignEnvironmentToViewerRequest` modifies the request. It's also important that the cache key contains the environment assigned by `assignEnvironmentToViewerRequest`. Otherwise, all users would be served the same environment. In our case the environment is assigned as a header named `x-hollowverse-requested-environment` and that header is whitelisted in CF settings to be a part of the cache key.
   3. If a cached response is available for the resulting cache key, CF returns that copy and the viewer request does not hit the origin server.
   4. If no cached response is available, CF will:
      a. execute the Lambda@Edge assigned to the `origin-request` stage, which is [`routeRequestToOrigin`](https://github.com/hollowverse/route-request/blob/master/src/routeRequestToOrigin.ts).
      b. `routeRequestToOrigin` reads the `x-hollowverse-requested-environment` header and modifies the request so that it is directed to the corresponding environment. If `x-hollowverse-requested-environment` was not whitelisted in CF settings, `routeRequestToOrigin` would not even see that header.
      c. Now that CF knows where to get the response, it goes ahead and fetches the page from the matching environment.
      d. An "environment" is an API Gateway API which proxies all requests to a regular Lambda function that serves the static files (`index.html`, scripts, stylesheets...).
      e. After fetching the response, CF executes the Lambda@Edge assigned to the `origin-response` stage. This is [`setHeadersOnOriginResponse`](https://github.com/hollowverse/route-request/blob/master/src/setHeadersOnOriginResponse.ts).
      f. `setHeadersOnOriginResponse`, as the name implies, sets some headers on the origin response, including `Set-Cookie` which allows the user to stay on the same environment for the lifetime of the cookie.
      g. CF takes the response _after_ it's modified by `setHeadersOnOriginResponse` and stores it in the edge cache so it's available for subsequent requests.

forabi commented:
The environment will serve `index.html` for all page requests.

`index.html` contains links to the scripts and stylesheets required to render the page.

Once the browsers fetches these files, React Router takes on from there. 

# NP pages

React Router reads the path in the address bar, e.g. `/Tom_Hanks` and makes a request to `api.hollowverse.com` to fetch the NP data for `Tom_Hanks`.

The API request looks like this:

```graphql
{
  notablePerson(slug:"Tom_Hanks") {
    name
    mainPhoto {
      url
      sourceUrl
    }
    editorialSummary {
      nodes {
        text
      }
      author
      lastUpdatedOn
    }
  }
}
```

The API response looks like this:

```json
{
  "data": {
    "notablePerson": {
      "name": "Tom Hanks",
      "mainPhoto": {
        "url": "https://photos.hollowverse.com/notable-people/Tom_Hanks.jpg",
        "sourceUrl": "https://commons.wikimedia.org/wiki/File:Tom_Hanks_2014.jpg"
      },
      "editorialSummary": {
        "nodes": [
          {
            "text": "Politically, Hanks falls squarely in the Democrat/liberal camp, like the vast majority of his Hollywood counterparts. He's stuck up for quite a few liberal causes like gay marriage and the environment, and has a large collection of electric and hybrid cars."
          },
          // ...
        ],
        "author": "Tom Kershaw",
        "lastUpdatedOn": "2012-04-05"
      }
    }
  }
}
```

We store the API data in Redux store and this change causes React to re-render the page with the full content.

The URL for the NP image is included in the API response. The images are hosted on a dedicated AWS S3 bucket. The bucket is put behind a CF distribution which is assigned to subdomain `photos.hollowverse.com`

## Search page
React Router reads the path `/search` from the address bar and renders the search page.

We use Algolia to provide search results. The search index is currently populated using an internal script but we will integrate Algolia with the API in the future so that any changes to the database are automatically reflected in Algolia's indices.

As soon as the user starts typing, we make a request to Algolia to fetch the search results. Names and tags (e.g. liberal, democrat) of NPs are used to find results matching the search query. We store Algolia's response in Redux store and this change causes React to re-render the page with the search results.

wholesomedev committed:
Switch to hash history

wholesomedev committed:
add .travils.yml

wholesomedev committed:
Adjust webpack variables

wholesomedev committed:
Fix build

wholesomedev committed:
Make title link to home

wholesomedev committed:
Create a path aware Link component

wholesomedev committed:
Rename visitor to user

wholesomedev committed:
Refactor all the things!

wholesomedev committed:
Render documentation correctly using `react-remarkable`

wholesomedev committed:
Fix styling of markdown rendering

wholesomedev committed:
Remove Redux and related dependencies

wholesomedev committed:
Clean up some dependencies from `package.json`

wholesomedev committed:
Allow some architecture links to not have documentation

wholesomedev committed:
Add AWS Route 53 component

wholesomedev committed:
Add hollowverseDotComCloudFront architecture

wholesomedev commented:
So as I was trying to document the information that you've shared @forabi, I started to realize that my choice in vis.js and the way I've laid out the architecture website and the code is not optimal.

I've decided to scrap what I built and start over using https://github.com/knsv/mermaid.

wholesomedev committed:
Add more components, links, and docs (now switch to mermaid!)

forabi commented:
Cool! I've heard about mermaid but I did not got a chance to use it.

wholesomedev committed:
Remove everything, just keep `.md` content

wholesomedev commented:
@forabi What you said about how the way things were setup could cause a maintenance burden really got to me that I deleted everything and just kept the `.md` files 😆

There's this [VS Code](https://marketplace.visualstudio.com/items?itemName=vstirbu.vscode-mermaid-preview) extension which renders Mermaid diagrams in the VS Code markdown preview.

I'm thinking for now we'll just update these markdown files in VS Code, and later we can figure out a way to convert the mermaid diagrams to images that we can embed into the markdown files and publish all of this stuff in a more consumable way online. Let me know if you have other ideas.

(I tried to use `mermaid.cli` but that gave me broken results.)

(By the way, I also tried using gitbook. I don't know why that thing is popular. It's not that good, to be honest.)

wholesomedev commented on hollowverseComCloudFront/hollowverseComCloudFront.md:
```diff
@@ -0,0 +1,37 @@
+
+When CloudFront receives the request for the page, it executes the Lambda@Edge assigned to the `viewer-request` stage. This is [`assignEnvironmentToViewerRequest`](https://github.com/hollowverse/route-request/blob/master/src/assignEnvironmentToViewerRequest.ts).
+
+`assignEnvironmentToViewerRequest` modifies the incoming user request and returns the modifies request [TK modifies the request based on what? Why does it modify the request?].
```
@forabi what does `assignEvironmentToViewerRequest` modify the request based on, and why does it modify the request?

forabi replied:
@wholesomedev As the name implies, it modifies the request to add an environment, specifically to add the `x-hollowverse-assigned-environment` header (whose value is a random environment picked based on weights if no `env` cookie exists in the request, or the value of `env` cookie if it exists and it is a valid environment).

wholesomedev commented on originServer/originServer.md:
````diff
@@ -0,0 +1,10 @@
+  servePage --> response(Response)
+```
+
+An "environment" is an API Gateway API which proxies all requests to a regular Lambda function that serves the static files (index.html, scripts, stylesheets...).
````
Can you tell me about the architecture of our origin servers so that we can document them in a detailed way?

(Feel free to pull down this branch and modify the content yourself)

forabi replied:
Well, there are no servers :smile:. It's all Lambdas.

Lambda do not have URLs. We cannot directly call a Lambda by visiting a URL in the browser. However, Lambdas can be launched in response to triggers.

A trigger can be
* a specific point in time set according to a schedule (like how `track-performance` is run everyday).
* a new file being added to S3 (like how `process-image` crops new images)
* an HTTP request incoming to API Gateway

So, API Gateway is a way to associate a URL with a Lambda. API Gateway gives us a URL like this:

```
https://nwfww37l0c.execute-api.us-east-1.amazonaws.com/master/Tom_Hanks
```

This is the API Gateway URL for the "API" wrapping our Lambdas (worth noting that an API in API Gateway is not limited to Lambda, but for us, all API does is proxying all requests the Lambda).

* `/log` is associated to the function [`serveLogEndpoint`](https://github.com/hollowverse/hollowverse/blob/fa81a9ddb39fa908f9eb8fe4524e46db486134fd/src/serveLogEndpoint.ts).
* Any other path (i.e. `/{proxy+}`) is associated to the function [`servePages`](https://github.com/hollowverse/hollowverse/blob/fa81a9ddb39fa908f9eb8fe4524e46db486134fd/src/servePages.ts).

The associations are defined [here in `serverless.yml`](https://github.com/hollowverse/hollowverse/blob/fa81a9ddb39fa908f9eb8fe4524e46db486134fd/serverless.yml#L30). Serverless takes care of creating the API and the Lambda functions and integrating the functions with the API.

Visiting a path like `/Tom_Hanks` will trigger an HTTP event in Lambda causing the associated function to execute. The function will respond with the response for `/Tom_Hanks`.

![screenshot from 2018-05-04 15-53-33](https://user-images.githubusercontent.com/1012761/39628976-54c6fc6e-4fb3-11e8-9edb-3a3e935d8319.png)

The Lambda@Edge `routeRequestToOrigin` routes environments to the API Gateway URLs (for example `beta` is routed to https://w5ch1jftqc.execute-api.us-east-1.amazonaws.com/beta and `master` is routed to https://nwfww37l0c.execute-api.us-east-1.amazonaws.com/master)

So the entire request lifecycle is like this (provided the request is not cached):

```
hollowverse.com/Tom_Hanks =>
Route 53 =>
(IP address of CloudFront) =>
assignEnvironmentToViewerRequest =>
routeRequestToOrigin =>
API Gateway URL (e.g. https://nwfww37l0c.execute-api.us-east-1.amazonaws.com/master) =>
servePages =>
setHeadersOnOriginRequest =>
CF edge cache =>
browser =>
(fetches scripts, stylesheets) =>
(makes request to api.hollowverse.com) =>
api.hollowverse.com =>
Route 53 =>
IP address of CloudFront distribution for the GraphQL API =>
API Gateway URL for the GraphQL API =>
Trigger Lambda associated with the API =>
Serve API response =>
(API response includes URL for NP image hosted on photos.hollowverse.com) =>
Render NP page with the API data and NP image
```

forabi replied:
I'll pull down the branch and add these later today.

wholesomedev commented on userRequest/userRequest.md:
```diff
@@ -0,0 +1,18 @@
+
+A user request to Hollowverse is handled as illustrated in this diagram.
+
+Hollowverse uses [AWS Route 53](./TK-link-to-configurations) for DNS management.
```
This is where I'd like us to provide links to our AWS configurations. Same for the Lambda functions and other services.

forabi committed:
Update CloudFront docs

forabi committed:
Describe origin request process in detail

forabi commented:
I've added a description of the origin request process and updated the hollowverse CloudFront documentation. Let me know what you think and if there's something that needs further explanation.

wholesomedev commented on originServer/originServer.md:
````diff
@@ -2,9 +2,45 @@
 ```
 
-An "environment" is an API Gateway API which proxies all requests to a regular Lambda function that serves the static files (index.html, scripts, stylesheets...).
+Lambda do not have URLs. We cannot directly call a Lambda by visiting a URL in the browser. However, Lambdas can be launched in response to triggers.
````
>Lambda do not have URLs. We cannot directly call a Lambda by visiting a URL in the browser.

@forabi Are you sure this is true? When I used Lambdas a couple of years ago, it was possible to hit a Lambda function directly from the client using AWS JS SDK. And the SDK obviously hit the Lambda using an HTTP URL.

forabi replied:
Why would the Lambda have to have a URL? Can't the SDK talk to the AWS API and say: hey, execute the function X for me and return whatever the execution result is?

wholesomedev replied:
It could. But as I remember, I didn't configure any AWS API Gateway at the time...

(I could be wrong)

forabi replied:
Sorry, I don't understand. You said you used the JS SDK to trigger the Lambda. This does not need a URL. It's similar to how I could go to the Lambda Dashboard and execute the Lambda with a test event.

wholesomedev committed:
I don't think the request has to come from CloudFront. The Origin Server can receive a request from any thing. So I renamed "CloudFront Origin Request" to just "Request".

I added the "What's the URL path?" decision shape to the flow chart, and added the different Lambdas based on that decision.

wholesomedev committed:
Though this is useful information, I feel it's not directly relevant to the Origin Server documentation.

wholesomedev committed:
I feel the image was not very helpful. I wasn't sure what part of the image was conveying the idea without stopping and really thinking about it.

wholesomedev committed:
Simplify text

wholesomedev committed:
Add documentation for after downloading index.html

wholesomedev commented:
@forabi I've added the sequence diagrams for what happens after downloading `index.html`. Please take a look and let me know if it makes sense (I still need to add the text documentation for the diagrams). After that I think we'll have documented all the big architectural components of our infrastructure.

I'd also like to add a deployment diagram and documentation. The relationship between GitHub, Travis, and Docker in our deployment.

Another thing we can add is code architecture documentation, things like Redux, React Router, RxJS, GraphQL, and data loader. But this might be hard to maintain because they evolve quickly...unless you feel there are pieces of that stuff that have more permanence? We can also wait to see which questions contributors/future team members may have around these things before we document them.

Also how easy/difficult is https://github.com/hollowverse/hollowverse/issues/451? I'm wondering if I should keep this PR open until https://github.com/hollowverse/hollowverse/issues/451 is closed or if we should merge with out it?

wholesomedev committed:
Convert `serveIndexHtml` to a sequence diagram

forabi commented:
> Also how easy/difficult is hollowverse/hollowverse#451? I'm wondering if I should keep this PR open until hollowverse/hollowverse#451 is closed or if we should merge with out it?

I think it's going to take some time to migrate all the repos and even then we would need to ensure all these secrets are only accessible to specific IAM roles/users before we can be confident we can expose our configurations publicly.

forabi commented on afterDownloadingIndexHtml/afterDownloadingIndexHtml.md:
```diff
@@ -0,0 +1,58 @@
+  participant api as api hollowverse com
+  participant aurora as AWS Aurora
+  participant photos as photos hollowverse com
+  participant s3 as AWS S3 bucket
```
photos.hollowverse.com is an alias for a CloudFront distrubition put in front of the S3 bucket. We may want to add CF in between.

forabi committed:
Fixes

forabi committed:
Add CF to sequence diagrams

forabi commented:
Looks good to me! Mermaid is really intuitive and easy to use! I've fixed a couple things and add CF to the sequence diagrams.

forabi commented:
> I'd also like to add a deployment diagram and documentation. The relationship between GitHub, Travis, and Docker in our deployment.

:+1: 

> Another thing we can add is code architecture documentation, things like Redux, React Router, RxJS, GraphQL, and data loader. But this might be hard to maintain because they evolve quickly...unless you feel there are pieces of that stuff that have more permanence? We can also wait to see which questions contributors/future team members may have around these things before we document them.

On the API side (GraphQL server + `dataloader`), things the architecture is unlikely to change. We can start documenting those now.

wholesomedev committed:
Restructure files in preparation of adding a build step

wholesomedev committed:
Add common clown config

wholesomedev committed:
Convert mermaid diagrams to PlantUML

wholesomedev committed:
Add PlantUML `svg` diagrams

wholesomedev committed:
Center diagrams

wholesomedev committed:
Add build tasks

wholesomedev committed:
Improve build process

wholesomedev committed:
Improve README.md and recommend PlantUML VSCode extension

wholesomedev committed:
Add more documentation

wholesomedev committed:
Add travis, update README, and docs
